CMakeLists.txt

```diff
@@ -441,15 +441,106 @@ endif()
+
+if(WOLFSSL_OPENSSH)
+    override_cache(WOLFSSL_CURVE25519 "yes")
+endif()
+
+if(WOLFSSL_CURVE25519)
+    if("${WOLFSSL_CURVE25519}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DCURVE25519_SMALL")
+        set(WOLFSSL_CURVE25519_SMALL "yes")
+    endif()
+
+    if("${WOLFSSL_CURVE25519}" STREQUAL "no128bit" OR WOLFSSL_32BIT)
+        list(APPEND WOLFSSL_DEFINITIONS "-DNO_CURVED25519_128BIT")
+    endif()
+
+    list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_CURVE25519")
+    set(WOLFSSL_FEMATH "yes")
+endif()
+
+# ED25519
+set(WOLFSSL_ED25519_SMALL "no")
+set(WOLFSSL_ED25519_HELP_STRING "Enable ED25519 (default: disabled)")
+set(WOLFSSL_ED25519 "no" CACHE STRING ${WOLFSSL_ED25519_HELP_STRING})
+set_property(CACHE WOLFSSL_ED25519 PROPERTY STRINGS "yes" "no" "small")
+
+if(WOLFSSL_OPENSSH)
+    override_cache(WOLFSSL_ED25519 "yes")
+endif()
+
+if(WOLFSSL_ED25519 AND NOT WOLFSSL_32BIT)
+    if("${WOLFSSL_ED25519}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DED25519_SMALL")
+        set(WOLFSSL_ED25519_SMALL "yes")
+        set(WOLFSSL_CURVE25519_SMALL "yes")
+    endif()
+
+    if(NOT WOLFSSL_SHA512)
+        message(FATAL_ERROR "cannot enable ed25519 without enabling sha512.")
+    endif()
+
+    set(WOLFSSL_FEMATH "yes")
+    set(WOLFSSL_GEMATH "yes")
+    list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_ED25519")
+endif()
+
+# CURVE448
+set(WOLFSSL_CURVE448_SMALL "no")
+set(WOLFSSL_CURVE448_HELP_STRING "Enable Curve448 (default: disabled)")
+set(WOLFSSL_CURVE448 "no" CACHE STRING ${WOLFSSL_CURVE448_HELP_STRING})
+set_property(CACHE WOLFSSL_CURVE448 PROPERTY STRINGS "yes" "no" "small")
+
+if(WOLFSSL_CURVE448)
+    if("${WOLFSSL_CURVE448}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DCURVE448_SMALL")
+        set(WOLFSSL_CURVE448_SMALL "yes")
+    endif()
+
+    if("${WOLFSSL_CURVE448}" STREQUAL "no128bit" OR WOLFSSL_32BIT)
+        list(APPEND WOLFSSL_DEFINITIONS "-DNO_CURVED448_128BIT")
+    endif()
+
+    list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_CURVE448")
+    set(WOLFSSL_FE448 "yes")
+endif()
+
+# ED448
+set(WOLFSSL_ED448_SMALL "no")
+set(WOLFSSL_ED448_HELP_STRING "Enable ED448 (default: disabled)")
+set(WOLFSSL_ED448 "no" CACHE STRING ${WOLFSSL_ED448_HELP_STRING})
+set_property(CACHE WOLFSSL_ED448 PROPERTY STRINGS "yes" "no" "small")
+
+if(WOLFSSL_ED448 AND NOT WOLFSSL_32BIT)
+    if("${WOLFSSL_ED448}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DED448_SMALL")
+        set(WOLFSSL_ED448_SMALL "yes")
+        set(WOLFSSL_CURVE448_SMALL "yes")
+    endif()
+
+    if(NOT WOLFSSL_SHA512)
+        message(FATAL_ERROR "cannot enable ed448 without enabling sha512.")
+    endif()
+
+    set(WOLFSSL_FE448 "yes")
+    set(WOLFSSL_GE448 "yes")
+    list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_ED448")
+
+    # EdDSA448 requires SHAKE256 which requires SHA-3
+    override_cache(WOLFSSL_SHAKE256 "yes")
```

Comment:
Encountered an error here as `WOLFSSL_SHAKE256` hasn't been set yet so it can't get the type. Maybe just need to move the `WOLFSSL_SHAKE256` above this?

reproducer
`rm -rf build && cmake -B build -DWOLFSSL_ED448=yes && make -C build`

```
CMake Error at cmake/functions.cmake:3 (set):
  set given invalid arguments for CACHE mode.
Call Stack (most recent call first):
  CMakeLists.txt:541 (override_cache)
```

Reply:
Nice catch. I didn't catch this because I had an existing CMakeCache.txt.

In the corresponding section of configure.ac for ED448, ENABLED_SHAKE256 is set before --enable-shake256 is parsed (using AC_ARG_ENABLE). I think the AC_ARG_ENABLE for shake256 should override the setting of ENABLED_SHAKE256. So, setting it to "yes" in the ED448 section effectively has no effect. Let me know if you agree; I might be missing something.

Reply:
Yep, configure.ac looks like it won't work in some cases. On x86_64/aarch64 SHA3 and SHAKE256 are enabled by default as , but other architectures it would fail. Should we fix configure.ac in the

Reply:
Something like https://github.com/elms/wolfssl/commit/6b628e9a65b90764dc387f25ec25ed00dfc2572f would work, not sure if it's the best solution for the configure.ac

Reply:
Yeah, I think we should fix the configure.ac and then replicate whatever logic we decide is best in CMake after (I'll modify this PR to do that). Perhaps another engineer with more experience with our autotools setup can review your code in the branch you linked. I'll take a look now too.

Reply:
Yeah, your change looks good to me. Thinking about this a bit deeper, it is somewhat inconvenient/error-prone how reliant configure.ac is on a specific ordering of all its statements. Makes me want to design the CMake system differently, rather than replicate the logic of configure.ac. It feels like we would need a dependency graph of all the enable-able features to guide that redesign, though. This graph exists implicitly in the ordering of configure.ac's different feature sections (the shake stuff depending on sha3 here is a perfect example). It would be nice if these dependencies were made explicit somehow in our CMake.

Reply:
ok I'll open a PR for the commit for `configure.ac`.

I agree that it's brittle. With so many feature flags I think it's a big lift to make them all explicit. In an ideal world that would be best.

Reply:
I mirrored the configure changes and pushed. Resolving this and assigning to another for review.